src/Messages/OWSMessageSender.m

```diff
@@ -594,6 +594,19 @@ - (void)deliverMessage:(TSOutgoingMessage *)message
 {
     TSThread *thread = message.thread;
 
+#ifdef DEBUG
+    // Assert that we're only sending sync messages when necessary
+    if ([message isKindOfClass:[OWSOutgoingSyncMessage class]]) {
+        __block BOOL hasSecondaryDevices = NO;
+        [self.storageManager.dbConnection readWithBlock:^(YapDatabaseReadTransaction *_Nonnull transaction) {
+            hasSecondaryDevices = [OWSDevice hasSecondaryDevicesWithTransaction:transaction];
+        }];
+        if (!hasSecondaryDevices) {
+            OWSFail(@"Sending sync message when we have no devices to sync to.");
+        }
```

Comment:
I started to go down this road, then realized that there's a race condition where a newly linked device immediately sends a group/contact sync request, before this device even knows about the device.

There's also an issue where it's possible to *not* persist the devices if the user leaves the device linking UI *just* as it completes.

So I'm going to put this on hold until the next release when we have more time to test it.